File: internal/diag/diag.go
package diag

import (
	"fmt"
	"strings"
	"yar/internal/token"
)

type Diagnostic struct {
	Pos     token.Position
	Message string
}

type List struct {
	items []Diagnostic
}

func (l *List) Add(pos token.Position, format string, args ...any) {
	l.items = append(l.items, Diagnostic{
		Pos:     pos,
		Message: fmt.Sprintf(format, args...),
	})
}

func (l *List) Append(other []Diagnostic) {
	l.items = append(l.items, other...)
}

func (l *List) Items() []Diagnostic {
	out := make([]Diagnostic, len(l.items))
	copy(out, l.items)
	return out
}

func (l *List) Empty() bool {
	return len(l.items) == 0
}

func Format(path string, diagnostics []Diagnostic) string {
	var b strings.Builder
	for i, d := range diagnostics {
		if i > 0 {
			b.WriteByte('\n')
		}
		diagPath := path
		if d.Pos.File != "" {
			diagPath = d.Pos.File
		}
		fmt.Fprintf(&b, "%s:%d:%d: %s", diagPath, d.Pos.Line, d.Pos.Column, d.Message)
	}
	return b.String()
}

File: crates/yar-compiler/src/diag.rs
use crate::{symbol::sanitize_diagnostic_message, token::Position};

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Diagnostic {
    pub pos: Position,
    pub message: String,
}

#[derive(Clone, Debug, Default)]
pub struct List {
    items: Vec<Diagnostic>,
}

impl List {
    pub fn add(&mut self, pos: Position, message: impl Into<String>) {
        self.items.push(Diagnostic {
            pos,
            message: sanitize_diagnostic_message(message.into()),
        });
    }

    pub fn append(&mut self, other: &[Diagnostic]) {
        self.items
            .extend(other.iter().cloned().map(|mut diagnostic| {
                diagnostic.message = sanitize_diagnostic_message(diagnostic.message);
                diagnostic
            }));
    }

    pub fn items(&self) -> Vec<Diagnostic> {
        self.items.clone()
    }

    pub fn is_empty(&self) -> bool {
        self.items.is_empty()
    }
}

pub fn format(path: &str, diagnostics: &[Diagnostic]) -> String {
    diagnostics
        .iter()
        .map(|diagnostic| {
            let diag_path = if diagnostic.pos.file.is_empty() {
                path
            } else {
                &diagnostic.pos.file
            };
            format!(
                "{}:{}:{}: {}",
                diag_path, diagnostic.pos.line, diagnostic.pos.column, diagnostic.message
            )
        })
        .collect::<Vec<_>>()
        .join("\n")
}
